### PR TITLE
Create extension point to connect the touring panel/Refactor LineupSidePanel

### PR DESCRIPTION
**Pull request by oltionchampari**
Closes datavisyn/tdp_core#257

Developer documentation:
https://wiki.datavisyn.io/tdp/tutorials/lineup-panel/lineup-side-panel
https://wiki.datavisyn.io/tdp/tutorials/lineup-panel/how-to-add-line-up-panel-tab
https://wiki.datavisyn.io/tdp/tutorials/lineup-panel/how-to-add-a-new-tdp-ranking-button

### Summary
- Refactored and extracted the LineupSidePanel class into several different files to make it more scalable 
- Added tabbing functionality to the side panel in able to have different content inside of it 
- Created extension point to add further tabs to the side panel

paneltab

**To register to the above extension point:**

```typescript
import { EP_TDP_CORE_LINEUP_PANEL_TAB, IPanelTabExtensionDesc } from 'tdp_core/src/extensions';

  registry.push(EP_TDP_CORE_LINEUP_PANEL_TAB, 'statisticalAnalysisPanel', function () {
    return System.import('./TouringPanel');
  }, <IPanelTabExtensionDesc>{
    cssClass: 'fa-calculator',
    title: 'Statistical Analysis',
    order: 10,
    width: '45em',
    shortcut: true
  });
```
Factory method name /entry point can be defined:
```typescript
factory:'<my factory method>'
```
If not defined the default factory method name is 
```typescript
create() 
```
### TODO

- [x] Test side panel in application where it is used in top mode

### CODE REVIEW

File: src/lineup/internal/LineUpPanelActions.ts [hunk 1]
```diff
@@ -58,39 +65,46 @@ export default class LineUpPanelActions extends EventHandler {
...
     }
-    this.node.classList.add('tdp-view-lineup');
-    this.collapse = options.enableSidePanel === 'top' || options.enableSidePanel === 'collapsed';
 
     this.init();
+    this.collapse = options.enableSidePanel === 'top' || options.enableSidePanel === 'collapsed';
+  }
+
+  removeTabButtonHighlighting() {
+    this.tabContainer.showDefault();
```
Comment (thinkh) [inline src/lineup/internal/LineUpPanelActions.ts]: It's a bit confusing that the header has to remove the highlight. Especially, because it is directly accessing the CSS classes of the tab. We might have to change the architecture of the tabs a bit. Maybe it helps if the `PanelNavButton` has a reference to his own tab or the `PanelTab` has an (optional) reference to the button that can activate it.

File: src/lineup/internal/LineUpPanelActions.ts [hunk 2]
```diff
@@ -114,14 +128,20 @@ export default class LineUpPanelActions extends EventHandler {
 
   set collapse(value: boolean) {
     this.node.classList.toggle('collapsed', value);
+    if (value && this.options.enableSidePanel !== 'top') {
+      this.removeTabButtonHighlighting();
+    }
```
Comment (thinkh) [inline src/lineup/internal/LineUpPanelActions.ts]: Why is this necessary for a side panel that is not top? Please add some comment to the code.

File: src/styles/_view_lineup.scss [hunk 3]
```diff
@@ -105,18 +105,19 @@ $lu_assets: '~lineupjs/src/assets';
...
   .lu-hierarchy .lu-search::before {
     height: 19px;
   }
 
-  .tdp-view-lineup.lu-side-panel {
-    flex: 0 0 auto;
-    position: relative;
-
+  //////////////////////
+  /////////////////////
```
Comment (thinkh) [inline src/styles/_view_lineup.scss]: Can this be removed?

File: src/lineup/internal/utils.ts [hunk 4]
```diff
@@ -7,12 +7,18 @@ import {IDataRow} from 'lineupjs';
 import {convertRow2MultiMap, IFormMultiMap, IFormRow} from '../../form';
 import {encodeParams} from 'phovea_core/src/ajax';
 
+
+
+export interface IAccessorFunc<T> {
+  (row: IDataRow): T;
+}
```
[the remark this answers is not in the document]
Comment (thinkh) [reply src/lineup/internal/utils.ts]: I think this one must be kept and the one in tourdino can import this one (from tdp_core). Please add some source code documentation here.

**Comment by oltionchampari** [comment]
Tested the side panel **top mode**  in patentbay  and it works and looks almost the same.
The only difference is that in patentbay a `margin-left:auto` is added which pushes the side panel to the right and with this branch it doesn't work anymore.
To fix it the css selector in patentbay needs to be adapted.
**Original:**
patent1


**Current branch:**
patent2

File: src/lineup/internal/panel/PanelButton.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    this.node.title = title;
+    this.node.addEventListener('click', (evt) => {
+      evt.stopPropagation();
+      evt.preventDefault();
+      this.highlight();
+      onClick();
+    });
+  }
+  private highlight() {
+    if (!this.tabNode.classList.contains('active')) {
```
[the remark this answers is not in the document]
Comment (oltionchampari) [reply src/lineup/internal/panel/PanelButton.ts]: Refactored implementation.

File: src/lineup/internal/panel/PanelButton.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      evt.preventDefault();
+      onClick();
+    });
+  }
+}
+
+export class PanelNavButton implements IPanelButton {
+  readonly node: HTMLElement;
+
+  constructor(parent: HTMLElement, private readonly tabNode: HTMLElement, title: string, linkClass: string, onClick: () => void) {
```
[the remark this answers is not in the document]
Comment (oltionchampari) [reply src/lineup/internal/panel/PanelButton.ts]: Added documentation.

**Comment by oltionchampari** [comment]
After a short discussion with Dominic, Christian and Michael  about the UI and the behavior of the nav tabs they suggested:

1. The **collapse/uncollapse** button should open the lineup config the first time it gets clicked  
 and the last open tab every other time.

2. There should be an icon for the default **lineup config tab** only in uncollapsed state though.

3. The **nav-tabs** should be on their own line after the default side-panel buttons.
 
Result : 

panel

File: src/lineup/internal/panel/PanelDownloadButton.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
+import {LocalDataProvider} from 'lineupjs';
+import {exportLogic} from '../export';
+import {IPanelButton} from './PanelButton';
+import i18n from 'phovea_core/src/i18n';
+export default class PanelDownloadButton implements IPanelButton {
```
[the remark this answers is not in the document]
Comment (oltionchampari) [reply src/lineup/internal/panel/PanelDownloadButton.ts]: Commented code.

File: src/lineup/internal/panel/PanelAddColumnButton.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
+import {SearchBox} from 'lineupjs';
+import {ISearchOption} from '../LineUpPanelActions';
+import {IPanelButton} from './PanelButton';
+import i18n from 'phovea_core/src/i18n';
+export default class PanelAddColumnButton implements IPanelButton {
```
[the remark this answers is not in the document]
Comment (oltionchampari) [reply src/lineup/internal/panel/PanelAddColumnButton.ts]: Commented code.

File: src/lineup/internal/panel/PanelAddColumnButton.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  readonly node: HTMLElement;
+
+  constructor(parent: HTMLElement, private readonly search: SearchBox<ISearchOption>) {
```
[the remark this answers is not in the document]
Comment (oltionchampari) [reply src/lineup/internal/panel/PanelAddColumnButton.ts]: Commented code.

File: src/lineup/internal/panel/PanelDownloadButton.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  readonly node: HTMLElement;
+
+  constructor(parent: HTMLElement, private provider: LocalDataProvider, isTopMode: boolean) {
```
[the remark this answers is not in the document]
Comment (oltionchampari) [reply src/lineup/internal/panel/PanelDownloadButton.ts]: Commented code.

File: src/lineup/internal/panel/PanelDownloadButton.ts [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+        const type = link.dataset.t;
+        const onlySelected = link.dataset.s === 's';
+        exportLogic(<any>type, onlySelected, this.provider).then(({content, mimeType, name}) => {
+          this.downloadFile(content, mimeType, name);
+        });
+      };
+    }
+  }
+
+  private downloadFile(content: BufferSource | Blob | string, mimeType: string, name: string) {
```
[the remark this answers is not in the document]
Comment (oltionchampari) [reply src/lineup/internal/panel/PanelDownloadButton.ts]: Commented code.

File: src/lineup/internal/panel/PanelHeader.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+  addButton(button: IPanelButton) {
+    this.buttons = [...this.buttons, button];
+    this.node.appendChild(button.node);
+  }
+
+  removeHighlighting() {
+    const highlightedButton = this.buttons.find((button) => button.node.classList.contains('active'));
+    if (highlightedButton) {
+      highlightedButton.node.classList.remove('active');
+    }
```
[the remark this answers is not in the document]
Comment (oltionchampari) [reply src/lineup/internal/panel/PanelHeader.ts]: Used bootstrap class to add the active styles `PanelNavButton`. We don't need to manually add the active class to the `PanelNavButton`.

File: src/lineup/internal/panel/PanelRankingButton.ts [hunk 13]
```diff
@@ -0,0 +1,20 @@
+import {IPanelButton} from './PanelButton';
+import {LocalDataProvider, Ranking} from 'lineupjs';
+
+export default class PanelRankingButton implements IPanelButton {
```
[the remark this answers is not in the document]
Comment (oltionchampari) [reply src/lineup/internal/panel/PanelRankingButton.ts]: Commented code.

**Comment by oltionchampari** [comment]
The current status looks like this:

panelTab

On a last refactoring round, changed the component structure to avoid confusing code.

File: src/lineup/internal/panel/PanelButton.ts [hunk 14]
```diff
@@ -33,27 +34,41 @@ export default class PanelButton implements IPanelButton {
...
-    if (!this.tabNode.classList.contains('active')) {
-      this.node.classList.add('active');
 
-    } else {
-      this.node.classList.remove('active');
+  /**
+   * When you click the shortcut button in collapsed mode focus on the navButton
+   */
+  setActive() {
+    const navButtons = Array.from(this.node.parentElement.children);
```
Comment (thinkh) [inline src/lineup/internal/panel/PanelButton.ts]: Avoid mutating the parent element!

File: src/lineup/internal/LineUpPanelActions.ts [hunk 15]
```diff
@@ -88,25 +88,31 @@ export default class LineUpPanelActions extends EventHandler {
...
+      const sidePanel = new SidePanelTab(this.node, this.searchBoxProvider.createSearchBox(), ctx, doc, options);
       this.panel = sidePanel.panel;
 
+      const listener = () => {
+        this.tabContainer.showTab(sidePanel);
+      };
       this.tabContainer = new PanelTabContainer(this.node);
-      this.tabContainer.addTab(sidePanel);
+      const lineupNavButton = new PanelNavButton(this.tabContainer.node, listener, options, true);
+      this.tabContainer.addTab(sidePanel, lineupNavButton);
```
Comment (thinkh) [inline src/lineup/internal/LineUpPanelActions.ts]: Do we need to create the PanelNavButton outside or could we create the button inside of `addTab`? Because the `PanelTab` (or here `SidePanelTab`) contains already the options.

File: src/lineup/internal/panel/PanelTabContainer.ts [hunk 16]
```diff
@@ -1,64 +1,124 @@
...
   }
 
-  addTab(tab: PanelTab) {
-    this.tabs = [...this.tabs, tab];
-    this.node.appendChild(tab.node);
+  /**
+   * Add nav-tab to the nav-tabs in the correct order.
+   * @param button PanelNavButton instance to add
+   */
+  addNav(nav: PanelNavButton) {
```
Comment (thinkh) [inline src/lineup/internal/panel/PanelTabContainer.ts]:
```suggestion
  addNav(nav: PanelNavButton) {
```

I would rename `addNav` to `addButton` as we are adding a button to the header.

File: src/lineup/internal/panel/PanelTabContainer.ts [hunk 17]
```diff
@@ -1,64 +1,124 @@
...
+    const sameOrderNav = this.navButtons.find((n) => n.order === nav.order); // find nav with same order
+    this.navButtons = [...this.navButtons, nav].sort((nav1, nav2) => nav1.order - nav2.order);
+    this.node.appendChild(nav.node);
+    const nextSibling = sameOrderNav || this.getNextSibling(this.navButtons, nav); // if same order property append new nav before it
+    if (nextSibling) {
+      this.node.insertBefore(nav.node, nextSibling.node);
+    } else {
+      this.node.appendChild(nav.node);
+    }
```
Comment (thinkh) [inline src/lineup/internal/panel/PanelTabContainer.ts]: We can shift the sorting to the extension point, as we only want to sort the panel tabs that are added by the extension point. That would simplify the code here.

File: src/lineup/internal/panel/PanelDownloadButtonContainer.ts [hunk 18]
```diff
@@ -2,10 +2,15 @@ import {LocalDataProvider} from 'lineupjs';
 import {exportLogic} from '../export';
 import {IPanelButton} from './PanelButton';
 import i18n from 'phovea_core/src/i18n';
-export default class PanelDownloadButton implements IPanelButton {
+
+
+/**
+ * A button dropdown to download selected/all rows of the ranking
+ */
+export default class PanelDownloadButtonContainer implements IPanelButton {
```
[the remark this answers is not in the document]
Comment (thinkh) [reply src/lineup/internal/panel/PanelDownloadButtonContainer.ts]: Please revert the renaming.

**Comment by oltionchampari** [comment]
I retested the side panel in _patentbay_ in top mode again and it works as expected.

File: src/lineup/internal/panel/PanelDownloadButton.ts [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+/**
+ * A button dropdown to download selected/all rows of the ranking
+ */
+export default class PanelDownloadButton implements IPanelButton {
+  readonly node: HTMLElement;
+
+  constructor(parent: HTMLElement, private provider: LocalDataProvider, isTopMode:boolean) {
+    this.node = parent.ownerDocument.createElement('div');
+    this.node.classList.add('btn-group', 'download-data-dropdown');
+    this.node.innerHTML = `
```
Comment (lehnerchristian) [inline src/lineup/internal/panel/PanelDownloadButton.ts]: why aren't you inserting the elements via `someHTMLNode.insertAdjacentHTML('afterbegin', '...')` [like you're doing here](https://github.com/datavisyn/tdp_core/blob/ochampari/10_tourdino-lineup-panel-button/src/lineup/internal/panel/PanelButton.ts#L54) or vice versa?

File: src/lineup/internal/LineUpPanelActions.ts [hunk 20]
```diff
@@ -217,78 +220,46 @@ export default class LineUpPanelActions extends EventHandler {
...
+          if (this.collapse) {
+            this.collapse = false; // expand side panel
+          }
+          this.tabContainer.showTab(tab);
 
-  protected saveRankingDialog(order: number[]) {
-    editDialog(null, (name, description, sec) => {
-      this.fire(LineUpPanelActions.EVENT_SAVE_NAMED_SET, order, name, description, sec);
+        } else {
+          plugin.load().then((p) => {
```
Comment (lehnerchristian) [inline src/lineup/internal/LineUpPanelActions.ts]: The interface for `factory` is `factory(parent: HTMLElement, provider: LocalDataProvider, desc: IRankingButtonExtensionDesc, idType: IDType, extraArgs: object): Promise<IScoreParam>;` according to `IPanelTabExtension`

so when you add the type `IPanelTabExtension` to p then the method is incompatible to the interface's definition. for example `tab.events` is passed as IDType

File: src/extensions.ts [hunk 21]
```diff
@@ -136,6 +146,40 @@ export interface IScoreColumnPatcherExtensionDesc extends IPluginDesc {
   load(): Promise<IPlugin & IScoreColumnPatcherExtension>;
 }
 
+export interface IPanelTabExtension {
+  desc: IPanelTabExtensionDesc;
+  factory(parent: HTMLElement, provider: LocalDataProvider, desc: IRankingButtonExtensionDesc, idType: IDType, extraArgs: object): Promise<IScoreParam>;
```
Comment (lehnerchristian) [inline src/extensions.ts]: should the `extraArgs` be optional?

File: src/lineup/internal/panel/PanelTab.ts [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+
+}
+
+/**
+ * The PanelTab creates a tab component that with can be toggled through the PanelNavButton
+ */
+export class PanelTab {
+
+  readonly node: HTMLElement;
+  readonly events: PanelTabEvents;
```
Comment (lehnerchristian) [inline src/lineup/internal/panel/PanelTab.ts]: usually we don't have instance variables like this and instead the class itself (`PanelTab` in this case) inherits from `EventHandler`). at least I haven't seen a usage like this yet

@thinkh do you have usages like this in Ordino?

File: src/lineup/internal/panel/PanelTab.ts [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+    return this.navButton;
+  }
+
+  getShortcutButton(): PanelButton {
+    const onClick = () => {
+      this.navButton.click();
+    };
+
+    // Note: `document.body` is used only for `parent.ownerDocument.createElement()` inside the button
+    return new PanelButton(document.body, this.options.title, 'fa ' + this.options.cssClass + ' shortcut-nav', onClick);
```
Comment (lehnerchristian) [inline src/lineup/internal/panel/PanelTab.ts]: so `document.body.ownerDocument` === `document`. I see that this parent HTMLElement passing is the pattern but that also does not make much sense I think :)

why not pass a reference to the real parent?

File: src/lineup/internal/panel/PanelTab.ts [hunk 24]
```diff
@@ -0,0 +1,134 @@
...
+
+  readonly panel: SidePanel | null;
+
+  /**
+   * @param parent The parent HTML DOM element
+   * @param search LineUp SearchBox
+   * @param ctx LineUp context
+   * @param doc Document
+   */
+  constructor(parent: HTMLElement, private readonly search: SearchBox<ISearchOption>, ctx: any, doc = document, options?: IPanelTabDesc) {
```
Comment (lehnerchristian) [inline src/lineup/internal/panel/PanelTab.ts]: the type of `ctx` is not any when a take a look at the `SidePanel` class

File: src/lineup/internal/panel/PanelTabContainer.ts [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+
+
+
+/**
+ * The header of the PanelTab
+ * Contains the PanelNavButtons that toggle the PanelTab
+ */
+class PanelTabHeader {
+  public node: HTMLElement;
+  public navButtons: PanelNavButton[] = [];
```
Comment (lehnerchristian) [inline src/lineup/internal/panel/PanelTabContainer.ts]: the PanelNavButtons are never pushed to this array, so it'll always be emtpy. if you push the buttons to this array, please make it private and readonly

File: src/lineup/internal/panel/SearchBoxProvider.ts [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+
+  get length(): number {
+    return this.searchBoxes.length;
+  }
+
+  /**
+   * Create a new LineUp SearchBox. The instance is added to the internal list and returned.
+   * @returns A new LineUp SearchBox instance
+   */
+  createSearchBox(options?: Partial<ISearchBoxOptions<ISearchOption>>): SearchBox<ISearchOption> {
```
Comment (lehnerchristian) [inline src/lineup/internal/panel/SearchBoxProvider.ts]: I just tested what happens when you pass `undefined` as a parameter to `Object.assign()`. it works, however I think it would be better to have the `options` parameter initialized as an empty object

**Comment by lehnerchristian** [comment]
I have just some minor questions and some comments on typings / unused variables

File: src/lineup/internal/LineUpPanelActions.ts [hunk 27]
```diff
@@ -114,14 +122,27 @@ export default class LineUpPanelActions extends EventHandler {
 
   set collapse(value: boolean) {
     this.node.classList.toggle('collapsed', value);
+
+    // When this.options.enableSidePanel === 'top' `this.collapsed=true` gets called. Attempting to open the PanelTab produces an error.
```
[the remark this answers is not in the document]
Comment (oltionchampari) [reply src/lineup/internal/LineUpPanelActions.ts]: When you have the `enableSidePanel='top'`  `this.collapse `gets set one time.
The problem is that i want the PanelTab to react accordingly when the SidePanel gets collapsed, apart from  when `enableSidePanel='top'` because there is no PanelTabs in top mode.
Running ` this.tabContainer.hideCurrentTab();` throws an error because` this.tabContainer` doesn't exist.

File: src/lineup/internal/panel/PanelDownloadButton.ts (hunk 19 above)
Comment (oltionchampari) [reply src/lineup/internal/panel/PanelDownloadButton.ts]: No special reason behind it. This can be unified.

File: src/lineup/internal/LineUpPanelActions.ts (hunk 20 above)
Comment (oltionchampari) [reply src/lineup/internal/LineUpPanelActions.ts]: Thank you, fixing in next commit.

File: src/extensions.ts (hunk 21 above)
Comment (oltionchampari) [reply src/extensions.ts]: It should.

File: src/lineup/internal/panel/PanelDownloadButton.ts (hunk 19 above)
Comment (oltionchampari) [reply src/lineup/internal/panel/PanelDownloadButton.ts]: No special reason, it can be unified.

File: src/lineup/internal/LineUpPanelActions.ts (hunk 20 above)
Comment (oltionchampari) [reply src/lineup/internal/LineUpPanelActions.ts]: If you could take a look again at the commit to confirm it looks better now.

File: src/lineup/internal/panel/PanelTab.ts (hunk 24 above)
Comment (thinkh) [reply src/lineup/internal/panel/PanelTab.ts]: Good point. I changed it.

File: src/lineup/internal/panel/PanelTab.ts (hunk 23 above)
Comment (thinkh) [reply src/lineup/internal/panel/PanelTab.ts]: Valid point. 😉 I changed it.

File: src/lineup/internal/panel/PanelTabContainer.ts (hunk 25 above)
Comment (thinkh) [reply src/lineup/internal/panel/PanelTabContainer.ts]: Thanks. We forgot to remove this in the last refactoring. I removed the variable now.

File: src/lineup/internal/panel/SearchBoxProvider.ts (hunk 26 above)
Comment (thinkh) [reply src/lineup/internal/panel/SearchBoxProvider.ts]: Good point! I changed it.

File: src/lineup/internal/LineUpPanelActions.ts [hunk 28]
```diff
@@ -114,14 +123,27 @@ export default class LineUpPanelActions extends EventHandler {
...
+    if (value && this.options.enableSidePanel !== 'top') {
+      this.tabContainer.hideCurrentTab(); // Hide the active PanelTab --> Inform its content to stop updating
+    } else if (this.options.enableSidePanel !== 'top') {
+      this.tabContainer.showCurrentTab(); // Show the last active PanelTab --> Inform its content to start updating again
+    }
```
[the remark this answers is not in the document]
Comment (thinkh) [reply src/lineup/internal/LineUpPanelActions.ts]: I just tried my solution and it still throws an error in the top mode. I suggest to add a `NullTabContainer` which exposes the public functions, but implements no operation. That way we would avoid having lots of if-else scattered in the code.

File: src/lineup/internal/LineUpPanelActions.ts (hunk 27 above)
[the remark this answers is not in the document]
Comment (thinkh) [reply src/lineup/internal/LineUpPanelActions.ts]: I changed the implementation and added a NullTabContainer, which has no functionality, for the top mode. This avoids these ugly if-else conditions.

File: src/lineup/internal/panel/PanelTab.ts (hunk 22 above)
Comment (thinkh) [reply src/lineup/internal/panel/PanelTab.ts]: You are right. Our intention was to pass an event object, so that the extension point can react to some events, and this way avoid to pass the whole tab to the extension point. Otherwise the extension point will have access to functionality it does not need or should not

Comment (thinkh) [reply src/lineup/internal/panel/PanelTab.ts]: We changed the implementation according to your suggestion and integrated the events into the panel tab.

**Comment by thinkh** [comment]
@lehnerchristian Thanks for your valuable review. We addressed your comments with the last few commits. As we agreed eariler today in a call, I'm gonna merge the this PR even though you haven't revised it again.
